src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            /// <summary>
+            /// Starts the wait thread.
+            /// </summary>
+            public void Start()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                waitThread.IsBackground = true;
+                waitThread.Start();
+            }
+
+            /// <summary>
+            /// Queues on the thread pool or executes directly a call to <see cref="UnregisterWait(object)"/>.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in an invalid <see cref="WaitHandle"/>
+            /// into <see cref="RegisteredWait.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the thread pool.
+            /// </remarks>
+            public void QueueOrExecuteUnregisterWait(RegisteredWait handle)
+            {
+                if (handle.IsBlocking)
+                {
+                    UnregisterWait(handle);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(UnregisterWait, handle);
+                }
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="state">The wait handle to unregister.</param>
+            private void UnregisterWait(object state)
+            {
+                RegisteredWait handle = (RegisteredWait)state;
+
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    Debug.Assert(Array.IndexOf(_registeredWaits, handle) != -1 || handle.IsUnregistered);
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_pendingRemoves, handle) == -1 && !handle.IsUnregistered)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                handle.SignalUserWaitHandle();
```

Comment:
If a signal is observed and a callback is queued, the event provided to Unregister should not be signaled until this callback runs. So the count needs to track how many callbacks are queued, not just how many are running at this moment. This is only necessary for callbacks queued when the wait handle is signaled, the timeout ones don't necessary have to be counted but I didn't think it's worth distinguishing them.